internal/tcp/tcp.go

```diff
@@ -228,349 +189,220 @@ var (
-		// no IPv6 header present on TCP packets received on the raw socket
-		ipHdrSize = 0
-	default:
-		return errors.New("unhandled AF family")
-	}
-	if err != nil {
-		return errors.Wrapf(err, "failed to create %s receive socket", af)
-	}
-
-	if err := bindToDevice(recvSocket, interfaceName); err != nil {
-		return errors.Wrapf(err, "failed to bind %s receive socket to interface %s", af, interfaceName)
-	}
+	logger.Info("TCP receiver starting...", zap.Int("AF", conn.AF))
 
 	// IP + TCP header, this channel is fed from the socket
 	in := make(chan Message, defines.ChannelInBufferSize)
 
 	go func() {
-		defer syscall.Close(recvSocket)
 		defer close(in)
 
-		rawPacket := make([]byte, maxTCPPacketSizeBytes)
 		for {
-			n, from, err := syscall.Recvfrom(recvSocket, rawPacket, 0)
+			pkt, err := conn.NextPacket()
 			// parent has closed the socket likely
 			if err != nil {
-				logger.Fatal("failed to receive from receiver socket",
-					zap.String("AF", af),
+				logger.Fatal("failed to receive packet from packet source",
 					zap.Error(err))
 			}
 			receiveTime = monoNow()
 
-			// IP + TCP header size
-			if n < ipHdrSize+tcpHdrSize {
-				logger.Error("n < ipHdrSize + tcpHdrSize",
-					zap.Int("ipHdrSize", ipHdrSize),
-					zap.Int("tcpHdrSize", tcpHdrSize))
-				continue
-			}
-
-			pkt, destinedToArachne := parsePkt(rawPacket[ipHdrSize:], listenPort)
-			if !destinedToArachne {
+			srcIP, DSCPv, err := parsePktIP(pkt)
+			if err != nil {
+				logger.Error("Error parsing packet IP layer", zap.Error(err), zap.Any("packet", pkt))
 				continue
 			}
 
-			var DSCPv DSCPValue
-			r := bytes.NewReader(rawPacket[1:2])
-			binary.Read(r, binary.BigEndian, &DSCPv)
-			if DSCPv < 0 {
-				logger.Warn("Received packet with invalid QoS DSCP value",
-					zap.Any("DSCP_value", DSCPv),
-					zap.Any("raw_packet", rawPacket))
+			tcpHeader, payloadTime, err := parsePktTCP(pkt)
+			if err != nil {
+				logger.Error("Error parsing packet TCP layer", zap.Error(err), zap.Any("packet", pkt))
 				continue
 			}
 
-			var fromAddr net.IP
-			switch af {
-			case "ip4":
-				fromAddr = net.IP((from.(*syscall.SockaddrInet4).Addr)[:])
-			case "ip6":
-				fromAddr = net.IP((from.(*syscall.SockaddrInet6).Addr)[:])
-			default:
-				logger.Fatal("unhandled AF family", zap.String("AF", af))
-			}
-
-			fromAddrStr := fromAddr.String()
 			switch {
-			case pkt.hasFlag(syn) && !pkt.hasFlag(ack):
+			case tcpHeader.SYN && !tcpHeader.ACK:
 				// Received SYN (Open port)
 				logger.Debug("Received",
 					zap.String("flag", "SYN"),
-					zap.String("src_address", fromAddrStr),
-					zap.Uint16("src_port", pkt.srcPort))
+					zap.Stringer("src_address", srcIP),
+					zap.Uint16("src_port", uint16(tcpHeader.SrcPort)))
 
 				// Replying with SYN+ACK to Arachne agent
-				srcPortRange := PortRange{pkt.srcPort, pkt.srcPort}
+				srcPortRange := PortRange{uint16(tcpHeader.SrcPort), uint16(tcpHeader.SrcPort)}
```

Comment:
Why cast here? Can we take care of this above (in `PortRange` for instance if needed)?